From 411ee7cd54fcbb4c8696b416509e4d8f622bbe6e Mon Sep 17 00:00:00 2001
From: Thao Luong <thao.luong.uw@rvc.renesas.com>
Date: Thu, 14 May 2015 19:36:16 +0700
Subject: [PATCH 1/1] omxaacdec: workaround for calculation wrong number of
 output frame when decode aacplus file

aacLC has sample per frame (spf) is 1024
aacplus has sample per frame is 2048
omxaacdec set fixed spf is 1024. This will lead to calculation
for number of output frames is wrong in case decode aacplus file.
Wrong number of output frames will make audiodecoder in base class confuse
input is 1 frame but output is 2 frames.
omxaacdec can not detect input is aacLC or aacplus
So this patch file will workaround for case decode aacplus

refs #69205
Signed-off-by: Thao Luong <thao.luong.uw@rvc.renesas.com>
---
 omx/gstomxaudiodec.c      |   12 +
 1 files changed, 12 insertions(+)

diff --git a/omx/gstomxaudiodec.c b/omx/gstomxaudiodec.c
index a6577b2..fcf1dda 100644
--- a/omx/gstomxaudiodec.c
+++ b/omx/gstomxaudiodec.c
@@ -32,6 +32,7 @@
 #include <string.h>
 
 #include "gstomxaudiodec.h"
+#include "gstomxaacdec.h"
 
 GST_DEBUG_CATEGORY_STATIC (gst_omx_audio_dec_debug_category);
 #define GST_CAT_DEFAULT gst_omx_audio_dec_debug_category
@@ -546,6 +547,17 @@ gst_omx_audio_dec_loop (GstOMXAudioDec * self)
         GST_WARNING_OBJECT (self, "Output buffer does not contain an integer "
             "number of input frames (frames: %d, spf: %d)", nframes, spf);
       nframes = (nframes + spf - 1) / spf;
+      if (GST_IS_OMX_AAC_DEC(self) && (nframes != 1))
+      {
+          /* aacLC has sample per frame (spf) is 1024
+           * aacplus has sample per frame is 2014
+           * omxaacdec set spf is 1024. This will lead to calculation
+           * for nframe is wrong in case decode aacplus file.
+           * Wrong nframe will make audiodecoder in base class confuse
+           * input is 1 frame but output is 2 frames. This code will 
+           * workaround for decode aacplus                           */
+           nframes = nframes / 2;
+      }
     }
 
     GST_BUFFER_TIMESTAMP (outbuf) =
-- 
2.4.0

